Userland/Libraries/LibWeb/Layout/TreeBuilder.cpp

```diff
@@ -339,6 +361,40 @@ ErrorOr<void> TreeBuilder::create_layout_tree(DOM::Node& dom_node, TreeBuilder::
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/rendering.html#button-layout
+    // If the element is an input element, or if it is a button element and its computed value for
+    // 'display' is not 'inline-grid', 'grid', 'inline-flex', or 'flex', then the element's box has
+    // a child anonymous button content box with the following behaviors:
+    if (is<HTML::HTMLButtonElement>(dom_node) && !(display == CSS::Display::from_short(CSS::Display::Short::InlineGrid) || display == CSS::Display::from_short(CSS::Display::Short::Grid) || display == CSS::Display::from_short(CSS::Display::Short::InlineFlex) || display == CSS::Display::from_short(CSS::Display::Short::Flex))) {
+        auto& parent = *dom_node.layout_node();
+
+        // If the box does not overflow in the vertical axis, then it is centered vertically.
+        auto table_computed_values = CSS::ComputedValues();
+        static_cast<CSS::MutableComputedValues&>(table_computed_values).set_display(CSS::Display::from_short(CSS::Display::Short::Table));
+        static_cast<CSS::MutableComputedValues&>(table_computed_values).set_height(CSS::Size::make_percentage(CSS::Percentage(100)));
+
+        auto cell_computed_values = CSS::ComputedValues();
+        static_cast<CSS::MutableComputedValues&>(cell_computed_values).set_display(CSS::Display { CSS::Display::Internal::TableCell });
+        static_cast<CSS::MutableComputedValues&>(cell_computed_values).set_vertical_align(CSS::VerticalAlign::Middle);
+
+        auto table_wrapper = parent.heap().template allocate_without_realm<BlockContainer>(parent.document(), nullptr, move(table_computed_values));
+        auto cell_wrapper = parent.heap().template allocate_without_realm<BlockContainer>(parent.document(), nullptr, move(cell_computed_values));
```

Comment:
The spec says to create an anonymous box with `display: flow-root`:

> The box is a [block-level](https://drafts.csswg.org/css-display/#block-level) [block container](https://drafts.csswg.org/css-display/#block-container) that establishes a new [block formatting context](https://drafts.csswg.org/css-display/#block-formatting-context) (i.e., ['display'](https://drafts.csswg.org/css2/#display-prop) is 'flow-root').

So I don't think we should be making tables here :)

Reply:
Hmm, sounds like we should be able to do that with flexbox and `align-self` + `justify-self` no?

Reply:
It would be possible to do with flex but the advantage of using tables is when the content overflows vertically then it loses its centering and is aligned to the top of the button container (as says the spec). If we were to use flex then you'd have to check somehow the height of the children, the height of the button container, and apply or not the flex layout.

Reply:
Yeah all right, let's start here then :)